timemachine/md/builders.py

```diff
@@ -77,24 +92,83 @@ def get_waters_to_delete():
         waters_to_delete = set()
         for idx in clashy_idxs:
             atom = all_atoms[idx]
-            waters_to_delete.add(atom.residue)
-            assert atom.residue.name == WATER_RESIDUE_NAME
+            if atom.residue.name == WATER_RESIDUE_NAME:
+                waters_to_delete.add(atom.residue)
         return waters_to_delete
 
     # First add back in the number of waters that are clashy. Then delete the clashy waters.
     # Done in this order so that additional waters being added will be at the boundaries, if added after deleting
     # addSolvent fills the void intended for the mols. Need to end up with the same number of waters as originally
     num_system_atoms = host_coords.shape[0]
     clashy_waters = get_waters_to_delete()
+    combined_templates = get_ion_residue_templates(modeller)
     # First add back in the number of waters that are clashy and we know we need to delete
-    modeller.addSolvent(host_ff, numAdded=len(clashy_waters), neutralize=False, model=sanitize_water_ff(water_ff))
+    modeller.addSolvent(
+        host_ff,
+        numAdded=len(clashy_waters),
+        neutralize=False,
+        model=sanitize_water_ff(water_ff),
+        residueTemplates=combined_templates,
+    )
     clashy_waters = get_waters_to_delete()
     modeller.delete(list(clashy_waters))
     assert num_system_atoms == modeller.getTopology().getNumAtoms()
 
 
+def solvate_modeller(
+    modeller: app.Modeller,
+    box: NDArray[np.float64],
+    ff: app.ForceField,
+    water_ff: str,
+    mols: list[Chem.Mol] | None,
+    ionic_concentration: float = 0.0,
+    neutralize: bool = False,
+):
+    """Utility function to handle the neutralization of adding mols"""
+    dummy_chain_id = "DUMMY"
+    if neutralize and mols is not None:
+        # Since we do not add the ligands to the OpenMM system, we add ions that emulate the net charge
+        # of the ligands so that `addSolvent` will neutralize the system correctly.
+        charges = [Chem.GetFormalCharge(mol) for mol in mols]
+        # If the charges are not the same for all mols, unable to neutralize the system effectively
+        # TBD: Decide if we want to weaken this assertion, most likely for charge hopping
+        assert all([charge == charges[0] for charge in charges])
+        charge = charges[0]
+        if charge != 0:
+            topology = app.Topology()
+            dummy_chain = topology.addChain(dummy_chain_id)
+            for _ in range(abs(charge)):
+                if charge < 0:
+                    res = topology.addResidue(CHLORINE_ION_RESIDUE, dummy_chain)
+                    topology.addAtom(CHLORINE_ION_RESIDUE, app.Element.getBySymbol("Cl"), res)
+                else:
+                    res = topology.addResidue(SODIUM_ION_RESIDUE, dummy_chain)
+                    topology.addAtom(SODIUM_ION_RESIDUE, app.Element.getBySymbol("Na"), res)
+            coords = np.zeros((topology.getNumAtoms(), 3)) * unit.angstroms
```

Comment:
Do we ever use the coords from the modeller directly for anything? (Maybe set to `np.nan` to be even more defensive?).

Reply:
Hm right, but as-is it sets them all to `(0,0,0)`?

Reply:
Hm... weird that it's even attempting to cast a `float` to an `int`. Anyways - I guess adding it to the origin is probably (?) okay - was just a bit surprised that `addSolvent` doesn't check for clashes or anything on the existing atoms. Will `addSolvent` add atoms/molecules that avoids the "origin" (0,0,0) in this case?